### ISSUE TITLE
Lambda Expressions

### ISSUE THREAD

**Issue by DavePearce** [Whiley/WhileyCompiler#1049]
There is a flaw in the way lambda "declarations" are typed.  Specifically, the following example illustrates:

```
type State is { int status }
type handler is method(State)|method(int)

type DomObject is &{
  method addEventListener(int[], handler),
  ...
}

method empty():
    skip

method register(DomObject obj):
    // Register identity
    obj->addEventListener("click",&(State s -> empty()))
```

The problem is that the expression `&(State s -> empty())` must have type `handler`, but the `Decl.Lambda` instance generated has type `method(State)`.

**Comment by DavePearce** [comment]
Can we somehow just use `LambdaAccess` for both, whereby in this case it refers to a `Decl.Lambda` rather than e.g. a `Decl.Function`.  I think this can work, though the whole issue with the `Binding` is problematic.

**Comment by DavePearce** [comment]
Fixed by #1132
